hazelcast/sql.py

```diff
@@ -0,0 +1,1604 @@
+
+        Returns:
+            Future[Iterator]:
+        """
+
+        def continuation(future):
+            response = future.result()
+
+            with self._lock:
+                if not response.row_metadata:
+                    # Can't get an iterator when we only have update count
+                    raise ValueError("This result contains only update count")
+
+                if self._iterator_requested:
+                    # Can't get an iterator when we already get one
+                    raise ValueError("Iterator can be requested only once")
+
+                self._iterator_requested = True
+
+                if should_get_blocking:
+                    iterator = _BlockingIterator(
+                        response.row_metadata,
+                        self._fetch_next_page,
+                        self._sql_service.deserialize_object,
+                    )
+                else:
+                    iterator = _FutureProducingIterator(
+                        response.row_metadata,
+                        self._fetch_next_page,
+                        self._sql_service.deserialize_object,
+                    )
+
+                # Pass the first page information to the iterator
+                iterator.on_next_page(response.row_page)
+                return iterator
+
+        return self._execute_response.continue_with(continuation)
+
+    def _fetch_next_page(self):
+        """Fetches the next page, if there is no fetch request
+        in-flight.
+
+        Returns:
+            Future[_SqlPage]:
+        """
+        with self._lock:
+            if self._fetch_future:
+                # A fetch request is already in-flight, return it.
+                return self._fetch_future
+
+            future = Future()
+            self._fetch_future = future
+
+            self._sql_service.fetch(
+                self._connection, self._query_id, self._cursor_buffer_size
+            ).add_done_callback(self._handle_fetch_response)
+
+            # Need to return future, not self._fetch_future, because through
+            # some unlucky timing, we might call _handle_fetch_response
+            # before returning, which could set self._fetch_future to
+            # None.
+            return future
+
+    def _handle_fetch_response(self, future):
+        """Handles the result of the fetch request, by either:
+
+        - setting it to exception, so that the future calls to
+          fetch fails immediately.
+        - setting it to next page, and setting self._fetch_future
+          to None so that the next fetch request might actually
+          try to fetch something from the server.
+
+        Args:
+            future (Future): The response from the server for
+            the fetch request.
+        """
+        try:
+            response = future.result()
+
+            response_error = self._handle_response_error(response["error"])
```

Comment:
Just a question, is `error` key always set in `response`?

Reply:
Yes, it comes from the codec. see sql_fetch_codec#decode_response